cfpb/open-source-checklist
# Scrub GHE references from your project's Git history

This script automates scrubbing the GHE URL from source code and git commits. It's mostly the same commands from the Dev Manual "Scrubbing Git History" page: [https://GHE/pages/CFPB/dev-manual/scrubbing-git.html](https://GHE/pages/CFPB/dev-manual/scrubbing-git.html)
## Additions
- `scrub` bash script which runs the commands to search git log for GHE references, then scrubs them using git filter-branch from both the source code and commit messages
- `.env_SAMPLE` - an environment variable is recommended for use with the `scrub` script to keep the ghe url from being committed to version control
## Changes
- Unlike the how-to in the Dev Manual, this script uses `git filter-branch` to update the commit messages instead of a manual interactive rebase. It's adapted from the "prepend" command here: https://davidwalsh.name/update-git-commit-messages with an additional flag to include tags in the updated history so we can carry tagged releases from GHE to github.com.
## Testing
1. Fork and clone a project from GHE locally[\* (see note below)](#note). For ease of testing I've created a repo called 'scrub-stories' there which you can clone and which won't matter if it breaks. [https://ghe/farmanc/scrub-stories](https://ghe/farmanc/scrub-stories)
2. Copy `.env_SAMPLE` to your project root.
3. Rename `.env_SAMPLE` to `.env`.
4. Edit `fake.ghe.url` to point to the real GHE url (do not include `https://`).
5. `cd` into the repo root to activate the new environment variable. You should receive a warning and then a confirmation message: 
   
   ![screenshot-env](https://cloud.githubusercontent.com/assets/702526/16022800/ac213c32-3187-11e6-9122-bfbb36d886a7.png)
6. In Terminal, from project root, run:
   
   ``` bash
   sh ./scrub.sh
   ```
7. Don't `git push` since it will affect collaborators and the git history of all branches!

<a id="note">**Note:**</a> This tool is meant to be used on any project we want to release on Github.com from GHE. So feel free to test it against any GHE repo. I've successfully tested it on CCDB-content and knowledgebase. Just be careful not to push to those remote repos - fork and clone new repos locally and delete when you're done, since we're rewriting Git history here.
## Review
- @virginiacc
- @contolini

[Preview this PR without the whitespace changes](?w=0)
## Screenshots

Success (after the first time you've run it, you should get this message on subsequent runs)
![screenshot-scrub-success](https://cloud.githubusercontent.com/assets/702526/16022177/8c6e52a6-3184-11e6-9d4a-2bd915d8fcc7.png)
## Todos
- tests (how?? ideas welcome)
- automate install + usage alongside other tasks including [license-exceptions](https://github.com/cfarm/license-exceptions)
## Checklist
